pkg/directorysync/client.go

```diff
@@ -0,0 +1,478 @@
+type UserEmail struct {
+	// Flag to indicate if this e-mail is primary.
+	Primary bool
+
+	// Directory User's e-mail.
+	Value string
+
+	// Type of e-mail (ex. work).
+	Type string
+}
+
+// User contains data about a provisioned Directory User.
+type User struct {
+	// The User's unique identifier.
+	ID string `json:"id"`
+
+	// The User's username.
+	Username string `json:"username"`
+
+	// The User's e-mails.
+	Emails []UserEmail `json:"emails"`
+
+	// The User's first name.
+	FirstName string `json:"first_name"`
+
+	// The User's last name.
+	LastName string `json:"last_name"`
+
+	// The User's raw attributes in raw encoded JSON.
+	RawAttributes json.RawMessage `json:"raw_attributes"`
+}
+
+// ListMetadata contains pagination options for Directory records.
+type ListMetadata struct {
+	// Pagination cursor to receive records before a provided ID.
+	Before string `json:"before"`
+
+	// Pagination cursor to receive records before a provided ID.
+	After string `json:"after"`
+}
+
+// ListUsersOpts contains the options to request provisioned Directory Users.
+type ListUsersOpts struct {
+	// Directory unique identifier.
+	Directory string
+
+	// Directory Group unique identifier.
+	Group string
+
+	// Maximum number of records to return.
+	Limit int
+
+	// Pagination cursor to receive records before a provided Directory ID.
+	Before string
+
+	// Pagination cursor to receive records after a provided Directory ID.
+	After string
+}
+
+// ListUsersResponse describes the response structure when requesting
+// provisioned Directory Users.
+type ListUsersResponse struct {
+	// List of provisioned Users.
+	Data []User `json:"data"`
+
+	// Cursor pagination options.
+	ListMetadata ListMetadata `json:"listMetadata"`
+}
+
+// ListUsers gets a list of provisioned Users for a Directory.
+func (c *Client) ListUsers(
+	ctx context.Context,
+	opts ListUsersOpts,
+) (ListUsersResponse, error) {
+	c.once.Do(c.init)
+
+	endpoint := fmt.Sprintf("%s/directory_users", c.Endpoint)
+	req, err := http.NewRequest(
+		http.MethodGet,
+		endpoint,
+		nil,
+	)
+	if err != nil {
+		return ListUsersResponse{}, err
+	}
+
+	req = req.WithContext(ctx)
+	req.Header.Set("Authorization", "Bearer "+c.APIKey)
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("User-Agent", "workos-go/"+workos.Version)
+
+	limit := ResponseLimit
+	if opts.Limit != 0 {
+		limit = opts.Limit
+	}
+	q := req.URL.Query()
+	if opts.Directory != "" {
+		q.Add("directory", opts.Directory)
+	}
+	if opts.Group != "" {
+		q.Add("group", opts.Group)
+	}
+	q.Add("before", opts.Before)
+	q.Add("after", opts.After)
+	q.Add("limit", strconv.Itoa(limit))
+	req.URL.RawQuery = q.Encode()
+
+	res, err := c.HTTPClient.Do(req)
+	if err != nil {
+		return ListUsersResponse{}, err
+	}
+	defer res.Body.Close()
+
+	if err = workos.TryGetHTTPError(res); err != nil {
+		return ListUsersResponse{}, err
+	}
+
+	var body ListUsersResponse
+	dec := json.NewDecoder(res.Body)
+	err = dec.Decode(&body)
```

Comment:
Can we default to a user provided JSON encoder like the SSO or Audit Trail packages.

https://github.com/workos-inc/workos-go/blob/master/pkg/sso/client.go#L56-L57

Reply:
Hmm, do we need a JSON encoder? The requests are all `GET`, and we're only passing query parameters for the resource IDs or pagination options.

I don't believe it's being used in SSO either [here](https://github.com/workos-inc/workos-go/blob/master/pkg/sso/client.go#L157).

Reply:
Ah, my bad, we are just decoding stuff here 👍